Sources/Bot/Services/DispatchService.swift

```diff
@@ -0,0 +1,198 @@
+            }
+            .observe(on: scheduler)
+            .startWithValues { pullRequests in
+                self.dispatchInitial(pullRequests: pullRequests)
+            }
+
+        gitHubEvents.events
+            .observe(on: scheduler)
+            .observeValues { [weak self] gitHubEvent in
+                switch gitHubEvent {
+                case let .pullRequest(event):
+                    self?.pullRequestDidChange(event: event)
+                case let .status(event):
+                    self?.statusChecksDidChange(event: event)
+                case .ping:
+                    break
+                }
+            }
+    }
+
+    private func dispatchInitial(pullRequests: [PullRequest]) {
+        let dispatchTable = Dictionary(grouping: pullRequests) { pullRequest in pullRequest.target.ref }
+        for (branch, pullRequestsForBranch) in dispatchTable {
+            makeMergeService(targetBranch: branch, scheduler: self.scheduler, initialPullRequests: pullRequestsForBranch)
+        }
+    }
+
+    private func pullRequestDidChange(event: PullRequestEvent) {
+        logger.log("📣 Pull Request did change \(event.pullRequestMetadata) with action `\(event.action)`")
+        let targetBranch = event.pullRequestMetadata.reference.target.ref
+        if let service = mergeServices[targetBranch] {
+            service.pullRequestChangesObserver.send(value: (event.pullRequestMetadata, event.action))
+        } else {
+            makeMergeService(targetBranch: targetBranch, scheduler: self.scheduler, initialPullRequests: [event.pullRequestMetadata.reference])
+        }
+    }
+
+    private func statusChecksDidChange(event: StatusEvent) {
+        // No way to know which MergeService this event is supposed to be for – isRelative(toBranch:) only checks for head branch not target so not useful here
+        // So we're sending it to all MergeServices, and they'll filter them themselves based on their own queues
+        for mergeServiceForBranch in mergeServices.values {
+            mergeServiceForBranch.statusChecksCompletionObserver.send(value: event)
+        }
+    }
+
+    @discardableResult
+    private func makeMergeService(targetBranch: String, scheduler: DateScheduler, initialPullRequests: [PullRequest] = []) -> MergeService {
+        let mergeService = MergeService(
+            targetBranch: targetBranch,
+            integrationLabel: integrationLabel,
+            topPriorityLabels: topPriorityLabels,
+            requiresAllStatusChecks: requiresAllStatusChecks,
+            statusChecksTimeout: statusChecksTimeout,
+            initialPullRequests: initialPullRequests,
+            logger: logger,
+            gitHubAPI: gitHubAPI,
+            scheduler: scheduler
+        )
+        self.mergeServices[targetBranch] = mergeService
+        self.healthcheck.startMonitoring(mergeServiceHealthcheck: mergeService.healthcheck)
+        mergeServiceLifecycleObserver?.send(value: .created(mergeService))
```

Comment:
maybe we should remove this out of this function so that this side effect is outside of the function?

Reply:
Not sure where would you move it to though? At what point in the code would you trigger this instead?

Reply:
Or were you just suggesting to just extract that in a new private function and call that private function from here?

Reply:
not from here but from the loop that calls this function:

```
for ... {
  let mergeService = makeMergeService(...)
  observeMergeServiceLifetime(mergeService)
}
```

not critical though, just surprised by the discardable result

Reply:
Ah right, I see what you mean.
But since we want to observe lifecycle _every time_ we create a new MergeService via `makeMergeService` it feels safer to add it inside the `makeMergeService` rather than making sure to always call `observeMergeServiceLifecycle` every time we call `makeMergeService`?